File: 1207-delete-nodes-and-return-forest/delete-nodes-and-return-forest.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        #s = set(to_delete)
        forest = []

        root = self.solve(root,to_delete,forest)

        if(root != None):
            forest.append(root)

        return forest

    def solve(self, root: Optional[TreeNode],s: List[int], forest: List[TreeNode]) -> TreeNode:
        if root == None:
            return None
        
        root.left = self.solve(root.left,s,forest)
        root.right = self.solve(root.right,s,forest)
        
        if root.val in s:

            if root.left:
                forest.append(root.left)
            if root.right:
                forest.append(root.right)
            
            return None
        
        return root

 

        

File: 1207-delete-nodes-and-return-forest/delete-nodes-and-return-forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        vector<TreeNode*> forest;

        root = solve(root, to_delete, forest);

        if (root != nullptr) {
            forest.push_back(root);
        }

        return forest;
    }
    TreeNode* solve(TreeNode* root, vector<int>& to_delete, vector<TreeNode*>& forest) {
        if (root == nullptr) {
            return nullptr;
        }

        root->left = solve(root->left, to_delete, forest);
        root->right = solve(root->right, to_delete, forest);

        if (find(to_delete.begin(), to_delete.end(), root->val) != to_delete.end()) {
            if (root->left != nullptr) {
                forest.push_back(root->left);
            }
            if (root->right != nullptr) {
                forest.push_back(root->right);
            }
            delete root;
            return nullptr;
        }

        return root;
    }
};